### PR TITLE
Fix bug with unions.

### LINKED ISSUE #926: Cannot form the union of more than two tables
### Description
I have a situation where I need to take the union over multiple tables (restriction or other methods will not work in this case). Taking the union over more than two tables throws an error.

### Reproducibility
Include:
- OS (MACOS)
- Python Version 3.8.8
- MySQL Deployment Strategy (remote)
- DataJoint Version 0.13.2

This is a trivial example that seems to reproduce the issue:
```python
@schema
class Params(dj.Lookup):
    definition = """
    params_id: tinyint unsigned
    """

Params.insert([[x] for x in range(3)])

params = [Params & {'params_id': x} for x in range(3)]

params[0] + params[1] + params[2]
```

Error stack:
```console
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
~/anaconda3/envs/neural-data-kit/lib/python3.8/site-packages/IPython/core/formatters.py in __call__(self, obj)
    700                 type_pprinters=self.type_printers,
    701                 deferred_pprinters=self.deferred_printers)
--> 702             printer.pretty(obj)
    703             printer.flush()
    704             return stream.getvalue()

~/anaconda3/envs/neural-data-kit/lib/python3.8/site-packages/IPython/lib/pretty.py in pretty(self, obj)
    392                         if cls is not object \
    393                                 and callable(cls.__dict__.get('__repr__')):
--> 394                             return _repr_pprint(obj, self, cycle)
    395 
    396             return _default_pprint(obj, self, cycle)

~/anaconda3/envs/neural-data-kit/lib/python3.8/site-packages/IPython/lib/pretty.py in _repr_pprint(obj, p, cycle)
    698     """A pprint that just redirects to the normal repr function."""
    699     # Find newlines and replace them with p.break_()
--> 700     output = repr(obj)
    701     lines = output.splitlines()
    702     with p.group():

~/anaconda3/envs/neural-data-kit/lib/python3.8/site-packages/datajoint/expression.py in __repr__(self)
    527         :rtype: str
    528         """
--> 529         return super().__repr__() if config['loglevel'].lower() == 'debug' else self.preview()
    530 
    531     def preview(self, limit=None, width=None):

~/anaconda3/envs/neural-data-kit/lib/python3.8/site-packages/datajoint/expression.py in preview(self, limit, width)
    531     def preview(self, limit=None, width=None):
    532         """ :return: a string of preview of the contents of the query. """
--> 533         return preview(self, limit, width)
    534 
    535     def _repr_html_(self):

~/anaconda3/envs/neural-data-kit/lib/python3.8/site-packages/datajoint/preview.py in preview(query_expression, limit, width)
     11     if width is None:
     12         width = config['display.width']
---> 13     tuples = rel.fetch(limit=limit + 1, format="array")
     14     has_more = len(tuples) > limit
     15     tuples = tuples[:limit]

~/anaconda3/envs/neural-data-kit/lib/python3.8/site-packages/datajoint/fetch.py in __call__(self, offset, limit, order_by, format, as_dict, squeeze, download_path, *attrs)
    190                 ret = return_values[0] if len(attrs) == 1 else return_values
    191         else:  # fetch all attributes as a numpy.record_array or pandas.DataFrame
--> 192             cur = self._expression.cursor(
    193                 as_dict=as_dict, limit=limit, offset=offset, order_by=order_by)
    194             heading = self._expression.heading

~/anaconda3/envs/neural-data-kit/lib/python3.8/site-packages/datajoint/expression.py in cursor(self, offset, limit, order_by, as_dict)
    511         if offset and limit is None:
    512             raise DataJointError('limit is required when offset is set')
--> 513         sql = self.make_sql()
    514         if order_by is not None:
    515             sql += ' ORDER BY ' + ', '.join(order_by)

~/anaconda3/envs/neural-data-kit/lib/python3.8/site-packages/datajoint/expression.py in make_sql(self, fields)
    111             distinct="DISTINCT " if distinct else "",
    112             fields=self.heading.as_sql(fields or self.heading.names),
--> 113             from_=self.from_clause(), where=self.where_clause())
    114 
    115     # --------- query operators -----------

~/anaconda3/envs/neural-data-kit/lib/python3.8/site-packages/datajoint/expression.py in from_clause(self)
     91             self._subquery_alias_count) if isinstance(src, QueryExpression)
     92             else src for src in self.support)
---> 93         clause = next(support)
     94         for s, left in zip(support, self._left):
     95             clause += ' NATURAL{left} JOIN {clause}'.format(

~/anaconda3/envs/neural-data-kit/lib/python3.8/site-packages/datajoint/expression.py in (.0)
     88 
     89     def from_clause(self):
---> 90         support = ('(' + src.make_sql() + ') as `$%x`' % next(
     91             self._subquery_alias_count) if isinstance(src, QueryExpression)
     92             else src for src in self.support)

~/anaconda3/envs/neural-data-kit/lib/python3.8/site-packages/datajoint/expression.py in make_sql(self)
    627             fields = arg1.primary_key
    628             return "({sql1}) UNION ({sql2})".format(
--> 629                 sql1=arg1.make_sql(fields),
    630                 sql2=arg2.make_sql(fields))
    631         # with secondary attributes, use union of left join with antijoin

TypeError: make_sql() takes 1 positional argument but 2 were given
```

### Expected Behavior
Forming the union over two elements of `params` works as expected with the same setup as described above.

Note: taking the union over more than two tables **does** work for python version 3.8.2 and datajoint version 0.12.6.

### PR DESCRIPTION
**Pull request by jverswijver**
Added handling for when union gets a union object inside its make_sql(), this allows you to do multiple unions.
Took the test from @guzman-raphael comments on the issue linked.

fix #926

### CODE REVIEW

**Comment by jverswijver** [comment]
Seems the test coverage api we use is experiencing a major outage, tests all pass locally. We can run the github actions when they are back up [(outage information)](https://status.coveralls.io/).

**Review by dimitri-yatsenko — APPROVED** [review]

File: datajoint/expression.py [hunk 1]
```diff
@@ -625,9 +625,15 @@ def make_sql(self):
...
-                sql2=arg2.make_sql(fields))
+            if isinstance(arg1, Union):
+                sql1 = arg1.make_sql()
+            else:
+                sql1 = arg1.make_sql(fields)
+            if isinstance(arg2, Union):
+                sql2 = arg2.make_sql()
+            else:
+                sql2 = arg2.make_sql(fields)
+            return "({sql1}) UNION ({sql2})".format(sql1=sql1, sql2=sql2)
```
Comment (guzman-raphael) [inline datajoint/expression.py:636]: Nice catch! :clap: 

Here is a small suggestion that is mindful of code coverage by use of ternary operators.

```suggestion
            return "({sql1}) UNION ({sql2})".format(
                sql1=arg1.make_sql() if isinstance(arg1, Union) else arg1.make_sql(fields),
                sql2=arg2.make_sql() if isinstance(arg2, Union) else arg2.make_sql(fields),
                )
```

**Review by guzman-raphael — CHANGES_REQUESTED** [review]
@jverswijver Nice work on this one man! :rocket:
Only small suggestions included but overall fix looks great.

File: tests/test_relational_operand.py [hunk 2]
```diff
@@ -505,3 +506,12 @@ def test_joins_with_aggregation():
...
         assert len(session_dates) == 1
+
+    @staticmethod
+    def test_union_multiple():
+        # https://github.com/datajoint/datajoint-python/issues/926
+        q1 = IJ & dict(j=2)
+        q2 = (IJ & dict(j=2, i=0)) + (IJ & dict(j=2, i=1)) + (IJ & dict(j=2, i=2))
+        x = set(zip(*q1.fetch('i', 'j')))
+        y = set(zip(*q2.fetch('i', 'j')))
+        assert_set_equal(x, y)
```
[the remark this answers is not in the document]
Comment (guzman-raphael) [reply tests/test_relational_operand.py:516]: @jverswijver Just tagged up on this point and looks like this test should have passed. Must still be something missing with this fix. Let me know once you've pinpointed the problem/solution. Thanks :handshake:

**Review by guzman-raphael — CHANGES_REQUESTED** [review]

**Comment by A-Baji** [comment]
It seems mysql [does not support nested unions](https://bugs.mysql.com/bug.php?id=3347) in 5.7. In 8.0 the issue is semi fixed where it allows left side nested unions, but throws a specific error for unsupported right side nested unions. However, there is a workaround where wrapping the query and all sub queries in a `SELECT * FROM ()` and then adding alias allows the query to execute properly.

File: datajoint/expression.py [hunk 3]
```diff
@@ -102,7 +102,7 @@ def from_clause(self):
         return clause
 
     def where_clause(self):
-        return '' if not self.restriction else ' WHERE(%s)' % ')AND('.join(
+        return '' if not self.restriction else ' WHERE (%s)' % ')AND('.join(
```
Comment (dimitri-yatsenko) [inline datajoint/expression.py:105]: we don't really need the extra space but no big deal

**Review by guzman-raphael — APPROVED** [review]
@jverswijver @A-Baji :rocket: Excellent job guys and great collab! :rocket: 

It is too bad that MySQL has such nuances in working with unions but glad we were able to workaround them to achieve full union nesting support.

**Comment by renniek** [comment]
Hi there, 

I am using datajoint version 0.13.7, and so I think this should be fixed based on what I've seen online for the different versions, but I am still getting this error when I try to do a union on more than 2 tables. Any advice? Thank you so much!

**Pull request merged**